src/extra/components/pulses.py

```diff
@@ -0,0 +1,422 @@
+
+        # Last chance for timeserver control data, likely
+        # empty for more recent data after 2020.
+        for source in data.control_sources:
+            m = cls._timeserver_control_re.match(source)
+            if m is not None:
+                timeserver_sources.add(m[0])
+
+        if len(timeserver_sources) > 1:
+            raise ValueError('multiple timeserver control sources found, '
+                             'please pass explicit one:\n' + ', '.join(
+                                sorted(timeserver_sources)))
+        elif timeserver_sources:
+            return timeserver_sources.pop()
+
+        raise ValueError('no timeserver found, please pass explicit source')
+
+    def _make_pulse_index(self, pulse_mask, with_pulse_id=True):
+        """Generate multi index for a given pulse mask."""
+
+        pulse_counts = pulse_mask.sum(axis=1)
+        train_ids = np.repeat(self._key.train_id_coordinates(), pulse_counts)
+
+        if with_pulse_id:
+            train_pulses = [mask.nonzero()[0] for mask in pulse_mask]
+            pulses_label = 'pulseId'
+        else:
+            train_pulses = [np.arange(count) for count in pulse_counts]
+            pulses_label = 'pulseNumber'
+
+        import pandas as pd
+        return pd.MultiIndex.from_arrays(
+            [train_ids, np.concatenate(train_pulses)],
+            names=['trainId', pulses_label])
+
+    def _get_pulse_mask(self, table):
+        """Generate pulse mask from a given bunch pattern table."""
+        raise NotImplementedError('_get_pulse_mask')
+
+    @property
+    def timeserver(self) -> SourceData:
+        """Used timeserver source."""
+        return self._source
+
+    @property
+    def bunch_pattern_table(self) -> KeyData:
+        """Used bunch pattern table key."""
+        return self._key
+
+    def select_trains(self, trains):
+        """Select a subset of trains in this data.
+
+        This method accepts the same type of arguments as
+        extra_data.DataCollection.select_trains()
+        """
+
+        res = copy(self)
+        res._source = self._source.select_trains(trains)
+        res._key = self._key.select_trains(trains)
+
+        return res
+
+    def get_pulse_counts(self, labelled=True):
+        """Get number of pulses per train.
+
+        Args:
+            labelled (bool, optional): Whether a labelled pandas Series
+                (default) or unlabelled numpy array is returned.
+
+        Returns:
+            (pandas.Series or numpy.ndarray) Number of pulses per train,
+                indexed by train ID if labelled is True.
+        """
+
+        counts = self._get_pulse_mask(self._key.ndarray()).sum(axis=1)
+
+        if labelled:
+            import pandas as pd
+            return pd.Series(
+                data=counts, dtype=np.int32,
+                index=self._key.train_id_coordinates())
+        else:
+            return counts
+
+    def get_pulse_id_row(self):
```

Comment:
Bikeshedding here, what do you think about `get_train_pulse_ids()` instead?

Reply:
I'm not sure... how about `peek_pulse_ids`?

Reply:
Yep, that sounds good too :+1: It's mostly the word `row` that threw me because I don't think we use that word anywhere else.